File: packages/ctf-core/src/references.ts
/**
 * Reference compression manager for CTF format
 */

import { JsonValue, ReferenceEntry } from './types.js';
import { isObject, isArray } from './utils.js';

export class ReferenceManager {
  private references: Map<string, number> = new Map();
  private reverseReferences: Map<number, string> = new Map();
  private valueCounts: Map<string, number> = new Map();
  private nextId: number = 1;

  /**
   * Build reference table from data
   */
  build(
    data: JsonValue,
    minOccurrences: number = 3,
    minLength: number = 5,
    minSavings: number = 10
  ): void {
    // Count all string values
    this.countStrings(data);

    // Create references for beneficial strings
    const entries: ReferenceEntry[] = [];

    for (const [value, count] of this.valueCounts) {
      if (count >= minOccurrences && value.length >= minLength) {
        const savings = this.calculateSavings(value, count);
        if (savings >= minSavings) {
          entries.push({ id: 0, value, occurrences: count, savings });
        }
      }
    }

    // Sort by savings (descending) and assign IDs
    entries.sort((a, b) => b.savings - a.savings);

    for (const entry of entries) {
      this.references.set(entry.value, this.nextId);
      this.reverseReferences.set(this.nextId, entry.value);
      this.nextId++;
    }
  }

  /**
   * Count all string occurrences in data
   */
  private countStrings(data: JsonValue): void {
    if (typeof data === 'string') {
      this.valueCounts.set(data, (this.valueCounts.get(data) || 0) + 1);
    } else if (isArray(data)) {
      for (const item of data) {
        this.countStrings(item);
      }
    } else if (isObject(data)) {
      for (const value of Object.values(data)) {
        this.countStrings(value);
      }
    }
  }

  /**
   * Calculate token savings from using a reference
   */
  private calculateSavings(value: string, count: number): number {
    // Without refs: value.length * count
    const withoutRefs = value.length * count;

    // With refs: 2 * count (for ^N) + value.length + 3 (for ^N=)
    const withRefs = 2 * count + value.length + 3;

    return withoutRefs - withRefs;
  }

  /**
   * Get reference ID for a value (if exists)
   */
  getReference(value: string): number | undefined {
    return this.references.get(value);
  }

  /**
   * Get value for a reference ID
   */
  getValue(id: number): string | undefined {
    return this.reverseReferences.get(id);
  }

  /**
   * Check if a value has a reference
   */
  hasReference(value: string): boolean {
    return this.references.has(value);
  }

  /**
   * Get all reference definitions as CTF lines
   */
  getDefinitions(): string[] {
    const lines: string[] = [];

    // Sort by ID
    const entries = Array.from(this.reverseReferences.entries()).sort((a, b) => a[0] - b[0]);

    for (const [id, value] of entries) {
      // Escape value if needed
      const escapedValue = this.escapeValue(value);
      lines.push(`^${id}=${escapedValue}`);
    }

    return lines;
  }

  /**
   * Escape a value for reference definition
   */
  private escapeValue(value: string): string {
    if (/[:|\t\n\r"^@]/.test(value) || value.includes('  ')) {
      return '"' + value.replace(/\\/g, '\\\\').replace(/"/g, '\\"') + '"';
    }
    return value;
  }

  /**
   * Get total number of references
   */
  get count(): number {
    return this.references.size;
  }

  /**
   * Clear all references
   */
  clear(): void {
    this.references.clear();
    this.reverseReferences.clear();
    this.valueCounts.clear();
    this.nextId = 1;
  }

  /**
   * Get statistics about references
   */
  getStats(): {
    totalReferences: number;
    totalSavings: number;
    avgSavingsPerRef: number;
  } {
    let totalSavings = 0;

    for (const [value, _id] of this.references) {
      const count = this.valueCounts.get(value) || 0;
      totalSavings += this.calculateSavings(value, count);
    }

    return {
      totalReferences: this.references.size,
      totalSavings,
      avgSavingsPerRef: this.references.size > 0 ? totalSavings / this.references.size : 0,
    };
  }
}

File: python/ctf/references.py
"""Reference compression for CTF format."""

from typing import Dict, Any, Optional
from ctf.utils import escape_string


class ReferenceManager:
    """Manages reference compression for repeated string values."""

    def __init__(self, min_length: int = 10, min_occurrences: int = 3) -> None:
        """
        Initialize reference manager.

        Args:
            min_length: Minimum string length to consider for references
            min_occurrences: Minimum occurrences to create a reference
        """
        self.min_length = min_length
        self.min_occurrences = min_occurrences
        self.value_counts: Dict[str, int] = {}
        self.references: Dict[str, str] = {}  # value -> ^N
        self.reverse_references: Dict[str, str] = {}  # ^N -> value
        self.next_ref_id = 1

    def analyze(self, value: Any) -> None:
        """
        Analyze value tree to count string occurrences.

        Args:
            value: Value to analyze (recursive)
        """
        if isinstance(value, str) and len(value) >= self.min_length:
            self.value_counts[value] = self.value_counts.get(value, 0) + 1
        elif isinstance(value, dict):
            for v in value.values():
                self.analyze(v)
        elif isinstance(value, list):
            for item in value:
                self.analyze(item)

    def build_references(self) -> None:
        """Build reference mappings for strings that meet criteria."""
        # Sort by savings potential (length * occurrences)
        candidates = [
            (value, count)
            for value, count in self.value_counts.items()
            if count >= self.min_occurrences
        ]
        candidates.sort(key=lambda x: len(x[0]) * x[1], reverse=True)

        # Create references
        for value, _ in candidates:
            ref = f"^{self.next_ref_id}"
            self.references[value] = ref
            self.reverse_references[ref] = value
            self.next_ref_id += 1

    def get_reference(self, value: str) -> Optional[str]:
        """
        Get reference for a value if it exists.

        Args:
            value: String value

        Returns:
            Reference (^N) or None
        """
        return self.references.get(value)

    def get_definitions(self) -> list[str]:
        """
        Get reference definitions as CTF lines.

        Returns:
            List of definition lines (^N="value")
        """
        definitions = []
        for ref, value in sorted(self.reverse_references.items(), key=lambda x: int(x[0][1:])):
            escaped = escape_string(value)
            definitions.append(f"{ref}={escaped}")
        return definitions

    def resolve_reference(self, ref: str) -> Optional[str]:
        """
        Resolve a reference to its value.

        Args:
            ref: Reference (^N)

        Returns:
            Original value or None if not found
        """
        return self.reverse_references.get(ref)

    def calculate_savings(self, value: str, count: int) -> int:
        """
        Calculate token savings for using a reference.

        Args:
            value: String value
            count: Number of occurrences

        Returns:
            Estimated token savings
        """
        # Original: count * len(value)
        # With ref: count * 2 (^N) + len(value) (definition)
        original = count * len(value)
        with_ref = count * 2 + len(value)
        return original - with_ref

    def is_reference(self, value: str) -> bool:
        """
        Check if a value is a reference.

        Args:
            value: Value to check

        Returns:
            True if value is a reference (^N)
        """
        return value.startswith("^") and value[1:].isdigit()
